pkg/controllers/elastic_jobset.go

```diff
@@ -0,0 +1,53 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"fmt"
+	"strconv"
+
+	batchv1 "k8s.io/api/batch/v1"
+
+	jobset "sigs.k8s.io/jobset/api/jobset/v1alpha2"
+)
+
+// jobsToDeleteDownScale gathers the excess jobs during a downscale
+// and deletes the jobs
+func jobsToDeleteDownScale(replicatedJobs []jobset.ReplicatedJob, replicatedJobStatus []jobset.ReplicatedJobStatus, jobItems []batchv1.Job) ([]*batchv1.Job, error) {
+	jobsToDelete := []*batchv1.Job{}
+	for _, replicatedJob := range replicatedJobs {
+		status := findReplicatedJobStatus(replicatedJobStatus, replicatedJob.Name)
+		countOfJobsToDelete := status.Ready - replicatedJob.Replicas
+		if countOfJobsToDelete > 0 {
+			jobsWeDeleted := 0
+			for _, val := range jobItems {
```

Comment:
Ii think it would be more optimal (although perhaps not provide a big perf improvement in practice) to store the jobs in a map of "replicatedJobName -> [slice of jobs for that replicatedJob]".  That way we don't need to iterate over every job in the jobset, for every replicatedJob we are downscaling. 

The O(jobs) cost of constructing the map is preferable to the O(replicatedJobs * jobs) cost of this approach.